feature: Buffer.lastIndexOf

dcposch opened the pull request:
Fixes #4604 

work done
---
* Added support for `Buffer.lastIndexOf` to match `Buffer.indexOf`
* Can search for a string, another Buffer, or a specific byte value, consistent with `Buffer.indexOf`
* For specific byte values, behavior is consistent with `Uint8Array.lastIndexOf`, which is now shadowed by `Buffer.lastIndexOf`, so existing code should continue to work
* Added test cases

work left to do
---
* Optimization. The implementation of reverse search in `string_search.cc` is naive and just uses a double for loop. Ideally we'd adapt `BoyerMooreSearch` to support reverse search, so that `lastIndexOf` will be equally fast as `indexOf`

Fishrock123 commented:
cc @trevnorris :)

mikeal commented:
Oh wow, I've been waiting years for this :)

This should make writing parsers a lot nicer :)

dcposch commented:
Cool. I added

performance
---
I've refactored so that both `lastIndexOf` and `indexOf` use the same fast algorithms: Boyer-Moore / Boyer-Moore-Horspool.

To do this with a minimum of messiness and without any code duplication, I expanded the `Vector<Char>` class that was already in `string_search.h` so that it can provide a *reversed view* onto a the underlying buffer. Then, the existing algorithms (Linear, Boyer-Moore, Boyer-Moore-Horspool) can be applied unmodified.

This works because `lastIndexOf(haystack, needle)` can be calculated from `indexOf(reverse(haystack), reverse(needle))`. Reversing the inputs is done via a lightweight view onto the original input buffer: it's not actually copying the buffers or doing anything slow like that.

testing
---
I added some additional test cases, to make sure that all of the search algorithms were being exercised. 

(*Background:* Under the hood, the existing `indexOf` starts with *Linear* search, then if that's too slow, switches to *Boyer-Moore-Horspool*, which requires only a quick precomputation but has O(n*m) worst-case complexity for finding a needle of size m in a haystack of size n. Finally, if that turns out to be too slow, it switches to *Boyer-Moore*, which requires more precomputation but has linear worst-case complexity.)

Surprisingly, the existing test cases never seem to exercise the Boyer-Moore fallback at all!

I added a test case that does go there.

If you want to reproduce the output below, compile with `DEBUG_STRING_SEARCH` defined and add a failing assertion to the bottom of `test-buffer-indexof.js` (otherwise the test runner swallows the output).

```
=== release test-buffer-indexof ===                                            
Path: parallel/test-buffer-indexof
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search BOYER-MOORE-HORSPOOL
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search BOYER-MOORE-HORSPOOL
forward search BOYER-MOORE-HORSPOOL
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search BOYER-MOORE-HORSPOOL
forward search BOYER-MOORE-HORSPOOL
forward search BOYER-MOORE-HORSPOOL
forward search BOYER-MOORE-HORSPOOL
forward search LINEAR
forward search LINEAR
forward search BOYER-MOORE-HORSPOOL
forward search BOYER-MOORE-HORSPOOL
forward search BOYER-MOORE-HORSPOOL
forward search BOYER-MOORE-HORSPOOL
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
forward search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search LINEAR
reverse search BOYER-MOORE-HORSPOOL
reverse search BOYER-MOORE-HORSPOOL
reverse search BOYER-MOORE-HORSPOOL
reverse search BOYER-MOORE
reverse search BOYER-MOORE-HORSPOOL
reverse search BOYER-MOORE
reverse search BOYER-MOORE-HORSPOOL
reverse search BOYER-MOORE
```

dcposch commented:
Also, this PR uses `memrchr` to search for a single byte value in a Buffer, back to front. Like `memchr`, it's probably a lot faster than just looping over bytes.

http://cebka.blogspot.com/2015/04/how-fast-is-your-memchr.html

**Unfortunately `memrchr` is GNU extension, not part of POSIX like `memchr`.**

1. Does Node have to compile in places where memrchr isn't available?
2. If so, should I try to pull in a polyfill, for example this one?
https://github.com/c9/node-gnu-tools/blob/master/grep-src/lib/memrchr.c

feross commented:
Thanks for tackling this PR, @dcposch!

dcposch commented:
Rebased on master

trevnorris commented on src/node_buffer.cc:
```diff
@@ -847,31 +882,25 @@ void IndexOfString(const FunctionCallbackInfo<Value>& args) {
 
   Local<String> needle = args[1].As<String>();
+  int64_t offset_i64 = args[2]->IntegerValue();
+  bool is_forward = args[4]->BooleanValue();
```
Faster to do `IsTrue()`

trevnorris commented on src/node_buffer.cc:
```diff
@@ -949,6 +983,8 @@ void IndexOfBuffer(const FunctionCallbackInfo<Value>& args) {
+  bool is_forward = args[4]->BooleanValue();
 
   if (buf_length > 0)
     CHECK_NE(buf_data, nullptr);
```
This is an artifact of old code. `SPREAD_ARG` already does this check. Feel free to remove these checks in applicable methods.

trevnorris commented on lib/buffer.js:
```diff
@@ -446,7 +446,46 @@ Buffer.prototype.compare = function compare(b) {
+  throw new TypeError('"val" argument must be string, number or Buffer');
+}
+
+function slowIndexOf(buffer, val, byteOffset, encoding, dir) {
```
style nit: functions in this file are two lines apart.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -949,6 +983,8 @@ void IndexOfBuffer(const FunctionCallbackInfo<Value>& args) {
   THROW_AND_RETURN_UNLESS_BUFFER(Environment::GetCurrent(args), args[0]);
```
Need to place `THROW_AND_RETURN_UNLESS_BUFFER(Environment::GetCurrent(args), args[1]);`. The value passed to `lastIndexOf()` could have had it's prototype messed with, resulting in the following:
```
$ ./node -e 'Buffer(4).lastIndexOf({__proto__: Buffer.prototype})'
node: ../src/node_buffer.cc:985: void node::Buffer::IndexOfBuffer(const FunctionCallbackInfo<v8::Value> &): Assertion `(args[1])->IsUint8Array()' failed.
Aborted (core dumped)
```
And even though they're messing around in a way they shouldn't, it still shouldn't be possible to cause an abort using the JS API.

dcposch replied:
sure, i can add that, but how did this work before? `IndexOfBuffer` was used by `indexOf` before this PR and didn't do the check

dcposch replied:
@trevnorris added

trevnorris replied:
Yup. It crashed before. Thanks for bringing it to my attention. :)

trevnorris commented:
Started to review, but have to step away. Will finish this tomorrow.

@dcposch Nice job on the inline comments. Makes the patch easier to follow. Especially for one this size.

dcposch commented:
@trevnorris thx. Fixed all the things you pointed out so far

trevnorris commented on lib/buffer.js:
```diff
@@ -446,7 +446,48 @@ Buffer.prototype.compare = function compare(b) {
+  } else if (byteOffset < -0x80000000) {
+    byteOffset = -0x80000000;
+  }
+  if (typeof byteOffset !== 'number' || isNaN(byteOffset)) {
```
This shouldn't be necessary. To follow `string.lastIndexOf()` the offset should be coerced to a primitive. So for example `'abcde'.lastIndexOf('c', [1]) === -1`. Basically the `byteOffset >>= 0` below should be enough.

dcposch replied:
I think we need this to match the behavior of `Buffer.indexOf`

* `buf.indexOf('foo')` searches the whole buffer, as does `buf.indexOf('foo', null)`, `buf.indexOf('foo', 'foo')`, etc
* `buf.indexOf('foo', 0)` searches starting from index 0, which also searches the whole buffer
* `buf.lastIndexOf('foo')` should def search the whole buffer, but
* `buf.lastIndexOf('foo', 0)` does a *reverse* search starting from index 0, so it only checks for a match at index 0

So a minimum, we have to special-case `undefined`

I think it's best if `buf.lastIndexOf('foo', null)`, `buf.lastIndexOf('foo', NaN)` etc match `buf.lastIndexOf('foo')` -- in other words,  they should search the whole buffer. That means they're NOT equivalent to `buf.lastIndexOf('foo', 0)`

trevnorris replied:
I agree. The operation is as simple as `offset = +offset`. This will coerce all `isNaN()` values to `NaN`, which can then be checked by `Number.isNaN()`. It will also coerce values like `[2]` to `2`, which is also how `String#lastIndexOf()` operates.

So any value that returns true for `Number.isNaN()` after the coercion is set to the default value. Though note this does exclude `null`. Which is the same way strings work. e.g. `'abc'.lastIndexOf('b', null) === -1`.

dcposch replied:
fixed. i added a test case to ensure `'abc'.lastIndexOf('b', null) === -1`

trevnorris commented on src/node_buffer.cc:
```diff
@@ -847,31 +882,25 @@ void IndexOfString(const FunctionCallbackInfo<Value>& args) {
   }
-
-  if (haystack_length < offset || needle_length + offset > haystack_length) {
+  size_t offset = static_cast<size_t>(opt_offset);
```
I'm a paranoid person. Mind adding a `CHECK_LT(offset, haystack_length);` just after this cast? Then also changing this conditional above from `opt_offset == -1` to `opt_offset <= -1`. That should help handle accidental under/overflow cases.

There's one other place where this is applicable. Will make a note when I get to it.

dcposch replied:
Sounds good, done

dcposch replied:
@trevnorris added `CHECK_LT` in the three places where I factored out `IndexOfOffset`

dcposch commented:
@trevnorris notice me senpai

Fishrock123 commented:
@dcposch some of us who work on this more *do* take the weekends off. ;)

trevnorris commented on src/string_search.h:
```diff
@@ -8,6 +8,15 @@
+#define DEBUG_TRACE(s) printf("%s search %s\n", \
+    subject.forward() ? "forward" : "reverse", s);
+#else
+#define DEBUG_TRACE(s)  // no-op
```
@bnoordhuis have any comments on this?

bnoordhuis replied:
I'd leave this out.

dcposch replied:
Removed. Note though that this was the only way I noticed some serious missing test coverage: the whole Boyer-Moore algorithm (nontrivial code, rarely exercised) was never reached by the unit tests before. (The unit tests previously only cover Boyer-Moore-Horspool, Linear, and Single-Char. After this PR they cover Boyer-Moore as well.)

trevnorris commented:
@dcposch Left few more comments. Now that the weekend is over will be more attentive. :)

dcposch commented:
@trevnorris fixed!

trevnorris commented:
Excellent work. CI: https://ci.nodejs.org/job/node-test-pull-request/1515/

dcposch commented:
I clicked Authorize, but it says
```
Access Denied
dcposch is missing the Overall/Read permission
```

rvagg commented:
Sorry @dcposch, we have CI in lockdown until we get our security releases out, you'll have to rely on collaborators to get you info on how the jobs have gone until it's opened back up again next week. / https://github.com/nodejs/node/issues/4857


There are compile errors on OSX:

```
In file included from ../src/node_buffer.cc:7:
../src/string_search.h:287:11: error: use of undeclared identifier 'memrchr'; did you mean 'memchr'?
    pos = memrchr(subject.start(), pattern_first_char, subj_len - index);
          ^~~~~~~
          memchr
/usr/include/string.h:70:7: note: 'memchr' declared here
void *memchr(const void *, int, size_t);
      ^
../src/node_buffer.cc:1061:11: error: use of undeclared identifier 'memrchr'; did you mean 'memchr'?
    ptr = memrchr(ts_obj_data, needle, offset + 1);
          ^~~~~~~
          memchr
/usr/include/string.h:70:7: note: 'memchr' declared here
void *memchr(const void *, int, size_t);
      ^
In file included from ../src/node_buffer.cc:7:
../src/string_search.h:252:18: error: use of undeclared identifier 'memrchr'
      void_pos = memrchr(subject.start(), search_byte, bytes_to_search);
                 ^
../src/string_search.h:308:10: note: in instantiation of function template specialization 'node::stringsearch::FindFirstCharacter<unsigned short>' requested here
  return FindFirstCharacter(search->pattern_, subject, index);
         ^
../src/string_search.h:107:22: note: in instantiation of member function 'node::stringsearch::StringSearch<unsigned short>::SingleCharSearch' requested here
        strategy_ = &SingleCharSearch;
                     ^
../src/string_search.h:607:22: note: in instantiation of member function 'node::stringsearch::StringSearch<unsigned short>::StringSearch' requested here
  StringSearch<Char> search(pattern);
                     ^
../src/string_search.h:636:36: note: in instantiation of function template specialization 'node::stringsearch::SearchString<unsigned short>' requested here
  size_t pos = node::stringsearch::SearchString(
                                   ^
../src/node_buffer.cc:928:16: note: in instantiation of function template specialization 'node::SearchString<unsigned short>' requested here
      result = SearchString(reinterpret_cast<const uint16_t*>(haystack),
               ^
In file included from ../src/node_buffer.cc:7:
../src/string_search.h:326:9: error: no matching function for call to 'FindFirstCharacter'
    i = FindFirstCharacter(pattern, subject, i);
        ^~~~~~~~~~~~~~~~~~
../src/string_search.h:110:20: note: in instantiation of member function 'node::stringsearch::StringSearch<unsigned short>::LinearSearch' requested here
      strategy_ = &LinearSearch;
                   ^
../src/string_search.h:607:22: note: in instantiation of member function 'node::stringsearch::StringSearch<unsigned short>::StringSearch' requested here
  StringSearch<Char> search(pattern);
                     ^
../src/string_search.h:636:36: note: in instantiation of function template specialization 'node::stringsearch::SearchString<unsigned short>' requested here
  size_t pos = node::stringsearch::SearchString(
                                   ^
../src/node_buffer.cc:928:16: note: in instantiation of function template specialization 'node::SearchString<unsigned short>' requested here
      result = SearchString(reinterpret_cast<const uint16_t*>(haystack),
               ^
../src/string_search.h:235:15: note: candidate template ignored: substitution failure [with Char = unsigned short]
inline size_t FindFirstCharacter(Vector<const Char> pattern,
              ^
../src/string_search.h:575:11: error: no matching function for call to 'FindFirstCharacter'
      i = FindFirstCharacter(pattern, subject, i);
          ^~~~~~~~~~~~~~~~~~
../src/string_search.h:113:18: note: in instantiation of member function 'node::stringsearch::StringSearch<unsigned short>::InitialSearch' requested here
    strategy_ = &InitialSearch;
                 ^
../src/string_search.h:607:22: note: in instantiation of member function 'node::stringsearch::StringSearch<unsigned short>::StringSearch' requested here
  StringSearch<Char> search(pattern);
                     ^
../src/string_search.h:636:36: note: in instantiation of function template specialization 'node::stringsearch::SearchString<unsigned short>' requested here
  size_t pos = node::stringsearch::SearchString(
                                   ^
../src/node_buffer.cc:928:16: note: in instantiation of function template specialization 'node::SearchString<unsigned short>' requested here
      result = SearchString(reinterpret_cast<const uint16_t*>(haystack),
               ^
../src/string_search.h:235:15: note: candidate template ignored: substitution failure [with Char = unsigned short]
inline size_t FindFirstCharacter(Vector<const Char> pattern,
              ^
5 errors generated.
make[2]: *** [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/obj.target/node/src/node_buffer.o] Error 1
make[2]: *** Waiting for unfinished jobs....
  c++ '-D_DARWIN_USE_64_BIT_INODE=1' '-DNODE_ARCH="x64"' '-DNODE_WANT_INTERNALS=1' '-DV8_DEPRECATION_WARNINGS=1' '-DHAVE_OPENSSL=1' '-DHAVE_DTRACE=1' '-D__POSIX__' '-DNODE_PLATFORM="darwin"' '-DHTTP_PARSER_STRICT=0' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' -I../src -I../tools/msvs/genfiles -I../deps/uv/src/ares -I/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/obj/gen -I../deps/v8 -I../deps/cares/include -I../deps/v8/include -I../deps/openssl/openssl/include -I../deps/zlib -I../deps/http_parser -I../deps/uv/include  -Os -gdwarf-2 -mmacosx-version-min=10.5 -arch x86_64 -Wall -Wendif-labels -W -Wno-unused-parameter -std=gnu++0x -fno-rtti -fno-exceptions -fno-threadsafe-statics -fno-strict-aliasing -MMD -MF /Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/.deps//Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/obj.target/node/src/tls_wrap.o.d.raw   -c -o /Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/obj.target/node/src/tls_wrap.o ../src/tls_wrap.cc
In file included from ../src/string_search.cc:1:
../src/string_search.h:287:11: error: use of undeclared identifier 'memrchr'; did you mean 'memchr'?
    pos = memrchr(subject.start(), pattern_first_char, subj_len - index);
          ^~~~~~~
          memchr
/usr/include/string.h:70:7: note: 'memchr' declared here
void *memchr(const void *, int, size_t);
      ^
1 error generated.
```

Windows

```
c:\workspace\node-compile-windows\label\win-vs2013\src\string_search.h(287): error C3861: 'memrchr': identifier not found (src\node_buffer.cc) [c:\workspace\node-compile-windows\label\win-vs2013\node.vcxproj]
```

smartos

```
In file included from ../src/node_buffer.cc:7:0:
../src/string_search.h: In function 'std::size_t node::stringsearch::FindFirstCharacter(node::stringsearch::Vector<const Char>, node::stringsearch::Vector<const Char>, std::size_t) [with Char = unsigned char; std::size_t = long unsigned int]':
../src/string_search.h:287:72: error: 'memrchr' was not declared in this scope
     pos = memrchr(subject.start(), pattern_first_char, subj_len - index);
                                                                        ^
../src/node_buffer.cc: In function 'void node::Buffer::IndexOfNumber(const v8::FunctionCallbackInfo<v8::Value>&)':
../src/node_buffer.cc:1061:50: error: 'memrchr' was not declared in this scope
     ptr = memrchr(ts_obj_data, needle, offset + 1);
                                                  ^
In file included from ../src/node_buffer.cc:7:0:
../src/string_search.h: In instantiation of 'size_t node::stringsearch::FindFirstCharacter(node::stringsearch::Vector<const Char>, node::stringsearch::Vector<const Char>, size_t) [with Char = short unsigned int; size_t = long unsigned int]':
../src/string_search.h:308:61:   required from 'static size_t node::stringsearch::StringSearch<Char>::SingleCharSearch(node::stringsearch::StringSearch<Char>*, node::stringsearch::Vector<const Char>, size_t) [with Char = short unsigned int; size_t = long unsigned int]'
../src/string_search.h:107:21:   required from 'node::stringsearch::StringSearch<Char>::StringSearch(node::stringsearch::Vector<const Char>) [with Char = short unsigned int]'
../src/string_search.h:607:36:   required from 'size_t node::stringsearch::SearchString(node::stringsearch::Vector<const Char>, node::stringsearch::Vector<const Char>, size_t) [with Char = short unsigned int; size_t = long unsigned int]'
../src/string_search.h:637:49:   required from 'size_t node::SearchString(const Char*, size_t, const Char*, size_t, size_t, bool) [with Char = short unsigned int; size_t = long unsigned int]'
../src/node_buffer.cc:933:39:   required from here
../src/string_search.h:252:71: error: 'memrchr' was not declared in this scope
       void_pos = memrchr(subject.start(), search_byte, bytes_to_search);
                                                                       ^
node.target.mk:157: recipe for target '/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos14-64/out/Release/obj.target/node/src/node_buffer.o' failed
make[2]: *** [/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos14-64/out/Release/obj.target/node/src/node_buffer.o] Error 1
```

the test also failed to run on armv8 but that was a jenkins problem as far as I can tell.

Looks like there's still quite a bit of work to do on cross-platform compat here.

dcposch commented:
@rvagg @trevnorris thanks!

Yeah this goes back to my first question at the top of the PR, about whether node needs to compile on systems that don't have `memrchr`. Looks like it does.

I added a fallback for those systems. LMK if the CI is happier now!

rvagg commented:
https://ci.nodejs.org/job/node-test-pull-request/1539/

sorry again you can't see that @dcposch, one of us will update with results when they come in

dcposch commented:
@rvagg thx!

rvagg commented:
All green except for some raspberry pi's but I was messing with the cluster while it was running and this accounts for 1/2 of the errors, the other 1/2 look like possible flaky tests. Running again @ https://ci.nodejs.org/job/node-test-pull-request/1540/

rvagg commented:
All green except for 2 failures on raspberry pi 2:

```
not ok 7 test-child-process-fork-regr-gh-2847.js
# TIMEOUT
  ---
  duration_ms: 120.61
  ...
```

Probably flaky: https://github.com/nodejs/node/issues/3635
Passed on the previous run so not a big concern.

```
not ok 138 test-http-regr-gh-2928.js
# TIMEOUT
# ..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++............................++..++++++++++++++++++++++++++++............................+.+.+++++++++++++++++++++++++++...........................++..++++++++++++++++++++++++++++............................+.+.+++++++++++++++++++++++++++...........................+.++++++++++++++++++++++++++++............................+.+++++++++++++++++++++++++++...........................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................++++++++++++++++++++++++++++............................+++++++++++++++++++++++++++...........................
  ---
  duration_ms: 120.63
  ...
```

Probably flaky: https://github.com/nodejs/node/issues/4830
But it failed on this and the previous run on rpi2, which troubles me a bit. @dcposch would you mind having a poke at the test and seeing if there's any way your changes could impact on that test?

Another full run for good measure (I may not be around to post the results of this): https://ci.nodejs.org/job/node-test-pull-request/1541/

Aside from that, looks like a job well done @dcposch! Will look to others to sign-off, I haven't actually reviewed the code.

rvagg commented:
same two, a bit of a concern:

```
not ok 7 test-child-process-fork-regr-gh-2847.js
# TIMEOUT
  ---
  duration_ms: 120.86
  ...

not ok 138 test-http-regr-gh-2928.js
# TIMEOUT
# ..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++++++..............................++++++++++++++++++++++++++..........................++++....++++++++++++++++++++++++++..........................+++...+.++++++++++++++++++++++++++..........................+++...+++++++++++++++++++++++++++...........................++..+.+++++++++++++++++++++++++++...........................++..++++++++++++++++++++++++++++............................++..+++++++++++++++++++++++++++...........................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++++++++++++++++.............................+++++++++++++++...............
  ---
  duration_ms: 120.63
  ...
```

trevnorris commented on lib/buffer.js:
```diff
@@ -446,7 +446,48 @@ Buffer.prototype.compare = function compare(b) {
+    return binding.indexOfNumber(buffer, val, byteOffset, dir);
+  }
+
+  throw new TypeError('"val" argument must be string, number or Buffer');
```
side note: more sure way here would be to pass all remaining values to C++, check if `node::Buffer::HasInstance()` or `args[1]->IsNumber()` then branch. Otherwise `env()->ThrowTypeError()`. Would get around potential proto munging that causes `instanceof` to incorrectly return true.

Don't bother making the change. Just making the note.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -941,17 +975,18 @@ void IndexOfString(const FunctionCallbackInfo<Value>& args) {
-  if (buf_length > 0)
-    CHECK_NE(buf_data, nullptr);
+  int64_t offset_i64 = args[2]->IntegerValue();
+  bool is_forward = args[4]->BooleanValue();
```
Should also be `IsTrue()` like above. No worries though, will fix that while landing.

dcposch replied:
fixed

trevnorris commented on src/node_buffer.cc:
```diff
@@ -1007,28 +1038,29 @@ void IndexOfBuffer(const FunctionCallbackInfo<Value>& args) {
   uint32_t needle = args[1]->Uint32Value();
   int64_t offset_i64 = args[2]->IntegerValue();
-  size_t offset;
+  bool is_forward = args[3]->BooleanValue();
```
ditto on fixing.

trevnorris commented:
All failing tests are not relevant to this PR.

Doing some quick profiling using `benchmarks/buffer/buffer-indexof.js` seems there's between 15-20% slow down for `Buffer#indexOf()` for `type=buffer encoding=utf8`. I'd like to at least know why that's happening before this lands. @dcposch Have cycles to investigate this? I can, but probably not until next week.

dcposch commented:
I did some benchmarking. Summary: https://docs.google.com/spreadsheets/d/1Lpxiagv8EnCS9B5F6ed0Z8cVJn91vej0M-OkF9_Qg8g/edit#gid=0
Click on the sheet tab to see comments.

**TLDR; looks like I added a constant overhead of about ~85 clock cycles per call to indexOf().**

The benchmark vary: some of them are equally fast before and after this PR, and others got slower. Looking closer, the heavy expensive calls to indexOf() are equally fast and the quick calls to indexOf() got noticeably slower. Running just the worst result (28% slow down): 

**Before (795eeb1): ~9.3m ops / second**
At 2.2GHz = ~235 clock cycles per indexOf()
```
$ ./node benchmark/buffers/buffer-indexof.js type=buffer encoding=undefined search=Alice
buffers/buffer-indexof.js search=Alice encoding=undefined type=buffer iter=1: 9249264.91467
$ ./node benchmark/buffers/buffer-indexof.js type=buffer encoding=undefined search=Alice
buffers/buffer-indexof.js search=Alice encoding=undefined type=buffer iter=1: 9303733.69067
```

**After (b67eb8c): ~6.8m ops / second**
At 2.2GHz = ~320 clock cycles per indexOf()
```
$ ./node benchmark/buffers/buffer-indexof.js type=buffer encoding=undefined search=Alice
buffers/buffer-indexof.js search=Alice encoding=undefined type=buffer iter=1: 6886913.37014
$ ./node benchmark/buffers/buffer-indexof.js type=buffer encoding=undefined search=Alice
buffers/buffer-indexof.js search=Alice encoding=undefined type=buffer iter=1: 6783969.18233
```

So we added about 85 cycles overhead for every call to indexOf(). How many of those are in JS vs in C++?
Stubbing out the C++ part in `node_buffer.cc` like this:
```c++
void IndexOfBuffer(const FunctionCallbackInfo<Value>& args) {
  if (true) return args.GetReturnValue().Set(0);
```

Before (JS only, stubbed C++): 120 cycles per call
After (JS only, stubbed C++): 145 cycles per call

#### Summary

We added ~85 cycles of overhead to every call to indexOf(). About 25 of those are in the JS wrapper and the other 60 are in C++.

I think the overhead comes from:
* The new function in JS, `bidirectionalIndexOf`
* The extra `direction` flag in JS and several of the C++ functions
* The couple of branches that check the direction flag

I don't know how we'd optimize the JS part, short of copy pasting `bidirectionalIndexOf` into `indexOf` and `lastIndexOf`. For the C++ part, I guess we could turn the `dir` into a template argument instead of a function argument so it makes separate object code for `indexOf` vs `lastIndexOf`.

Dunno if we care.

trevnorris commented:
@dcposch Haven't forgotten about this. Will get to it next week.

dcposch commented:
@trevnorris thx

dcposch commented:
@trevnorris want me to squash and rebase?

trevnorris commented:
@dcposch sounds good.

dcposch commented:
@trevnorris done

jasnell commented:
LGTM if @trevnorris is happy

trevnorris commented:
LGTM

jasnell commented:
New CI, just to be safe: https://ci.nodejs.org/job/node-test-pull-request/1808/

jasnell commented:
@dcposch @trevnorris ... is this ready to land?

trevnorris commented:
@jasnell I'd like @bnoordhuis to take a quick look if he has the time. If not then I'm alright that this lands.

dcposch commented:
Combine doc and feature PRs here at @Qard 's request. The new docs were LGTMd by @jasnell in this other PR: #4933

feross commented:
ping @bnoordhuis

jasnell commented:
@dcposch ... looks like this needs to be rebased and updated.
@bnoordhuis ... ping

dcposch commented:
@jasnell @bnoordhuis done

dcposch commented:
By the way, I'm getting the following linter error--not just on this but on the latest master (https://github.com/nodejs/node/commit/6a3d38f00f116cbe51cd006ee558ca773522e6f5) as well.

Is master currently borked?

```
/home/dc/node/benchmark/http_simple_auto.js
  16:13  error  Strings must use singlequote  quotes
  35:19  error  Strings must use singlequote  quotes
  44:12  error  'i' is not defined            no-undef
  44:19  error  'i' is not defined            no-undef
  44:26  error  'i' is not defined            no-undef
  45:25  error  'i' is not defined            no-undef
  77:10  error  'i' is not defined            no-undef
  77:37  error  'i' is not defined            no-undef
  77:46  error  'i' is not defined            no-undef
  78:28  error  'i' is not defined            no-undef
  78:38  error  'i' is not defined            no-undef

✖ 11 problems (11 errors, 0 warnings)
```

jasnell commented:
Should be good to go now. This was my bad (https://github.com/nodejs/node/commit/cde81b6b9f3d39629ccacae5f9e82ba2527c4f29)

dcposch commented:
@jasnell cool, rebased

jasnell commented:
Awesome, this LGTM but still want to get @bnoordhuis' final review

bnoordhuis commented on lib/buffer.js:
```diff
@@ -502,7 +502,48 @@ Buffer.prototype.compare = function compare(b) {
+  } else if (byteOffset < -0x80000000) {
+    byteOffset = -0x80000000;
+  }
+  byteOffset = +byteOffset; // coerce to Number
```
Stylistic issue: comments (both in JS and C++) should be capitalized and punctuated, and, if they're on the same line, offset by two spaces.

dcposch replied:
fixed

bnoordhuis commented on src/string_search.h:
```diff
@@ -21,60 +21,35 @@ T Max(T a, T b) {
  public:
-  Vector(T* data, size_t length) : start_(data), length_(length) {
+  Vector(T* data, size_t length, bool isForward)
+      : start_(data), length_(length), isForward_(isForward) {
```
Stylistic issue: the data member's name should be `is_forward_`.

dcposch replied:
fixed

bnoordhuis commented on src/string_search.h:
```diff
@@ -21,60 +21,35 @@ T Max(T a, T b) {
-      }
-    }
-    return true;
+    return start_[isForward_ ? index : (length_ - index - 1)];
```
I wonder if adding a branch here impacts performance.

bnoordhuis commented on src/string_search.h:
```diff
@@ -165,42 +126,42 @@ class StringSearch : private StringSearchBase {
-                      sizeof(PatternChar) == sizeof(uint16_t),
-                  "sizeof(PatternChar) == sizeof(uint16_t) || sizeof(uint8_t)");
+    static_assert(sizeof(Char) == sizeof(uint8_t) ||
+                      sizeof(Char) == sizeof(uint16_t),
```
Can you line up the arguments while you're here?

dcposch replied:
done

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+// Falls back to a vanilla for loop on systems which don't
+inline const void* MemrchrFill(const void* haystack, uint8_t needle,
+                               size_t haystack_len) {
+#ifdef _GNU_SOURCE
```
I don't think we compile with `-D_GNU_SOURCE` so this is probably dead code.

dcposch replied:
`_GNU_SOURCE` is included by default when compiling against glibc, which I think we are on Mac and Linux.

http://stackoverflow.com/questions/11670581/why-is-gnu-source-defined-by-default-and-how-to-turn-it-off

I just tested it & confirmed it's defined when I run `make`. I'm using Ubuntu 15.10.

I think we want to keep using the built-in `memrchr` when available because it's faster than a simple for loop: http://cebka.blogspot.com/2015/04/how-fast-is-your-memchr.html

bnoordhuis replied:
`_GNU_SOURCE` is only auto-defined on Linux.  That's just as well for OS X because I don't believe it has a `memrchr()` function (at least, 10.7 doesn't and that's our baseline) but FreeBSD and OpenBSD do.

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+#ifdef _GNU_SOURCE
+  return memrchr(haystack, needle, haystack_len);
+#else
+  const uint8_t *haystack8 = reinterpret_cast<const uint8_t*>(haystack);
```
Style: `const uint8_t*` (star leans left.)  Also, just `static_cast` should work here.

dcposch replied:
Sure, I'll change it. Just noting I've always seen the style guides and linters enforcing the opposite -- star goes to the right. That's because:

**a, b, and c are all `int`s**
```
int a, b, c;
```


**a is an `int*`, b and c are `int`s**
```
int* a, b, c;
```

**a, b, and c are all `int*`**
```
int *a, *b, *c;
```

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+  return memrchr(haystack, needle, haystack_len);
+#else
+  const uint8_t *haystack8 = reinterpret_cast<const uint8_t*>(haystack);
+  for (size_t i = haystack_len - 1; i != (size_t)-1; i--) {
```
`static_cast<size_t>(-1)`

dcposch replied:
fixed

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+    size_t bytes_to_search;
+    const void* void_pos;
+    if (subject.forward()) {
+      bytes_to_search = (max_n - pos) * sizeof(Char);
```
Theoretically prone to wraparound.

dcposch replied:
do you want me to add asserts?

bnoordhuis replied:
That would be good, yes.

dcposch replied:
@bnoordhuis done, let me know if that looks better

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+      bytes_to_search = (subject.length() - pos) * sizeof(Char);
+      void_pos = MemrchrFill(subject.start(), search_byte, bytes_to_search);
+    }
+    const Char* char_pos = reinterpret_cast<const Char*>(void_pos);
```
`static_cast<const Char*>(...)` should be sufficient here.

dcposch replied:
fixed

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+
+    // Then, for each match, verify that the full two bytes match pattern[0]
+    char_pos = AlignDown(char_pos, sizeof(Char));
+    size_t raw_pos = static_cast<size_t>(char_pos - subject.start());
```
Is `char_pos <= subject.start()` when searching backwards?

dcposch replied:
No `char_pos >= subject.start()`, regardless of whether we're searching forward or backward.

We're looking for `pattern` in `subject`. Every `char_pos` value is a potential match and should always be within the bounds of `subject`

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+  } else {
+    pos = MemrchrFill(subject.start(), pattern_first_char, subj_len - index);
+  }
+  const uint8_t* char_pos = reinterpret_cast<const uint8_t*>(pos);
```
`static_cast`

dcposch replied:
fixed

bnoordhuis commented on src/string_search.h:
```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
-                    pattern_length - 1)) {
-      return i - 1;
+    bool matches = true;
+    for (size_t j = 1; j < pattern_length; j++) {
```
You're skipping over the first character here?

bnoordhuis commented on src/string_search.h:
```diff
@@ -655,16 +632,33 @@ size_t SearchString(Vector<const SubjectChar> subject,
+  size_t diff = haystack_length - needle_length;
+  size_t relative_start_index = is_forward ? start_index
+      : (diff < start_index) ? 0
+      : (diff - start_index);
```
It's arguably more readable to write it as:
```c++
size_t relative_start_index;
if (is_forward)
  relative_start_index = start_index;
else
  relative_start_index = /* ... */;
```

dcposch replied:
yeah i agree. fixed

dcposch commented:
@bnoordhuis thanks. I made all the style changes you pointed out, and I've also gone thru and capitalized and punctuated all the comments. Let me know if that looks better.

dcposch commented:
@bnoordhuis re: performance, I ran the `buffer-indexof` benchmark before and after this change.

Here's a spreadsheet summarizing the results:
https://docs.google.com/spreadsheets/d/1lIEwi3VKM3bcqrIAfKSnJf-sOfrDMga3W94l-lxqe8M/edit?usp=sharing

Looks like `buffer.indexOf` got about 6% slower. Don't know if that's a problem.

trevnorris commented on src/string_search.h:
```diff
@@ -274,111 +230,138 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
+    if (subject.forward()) {
+      // Assert that bytes_to_search won't overflow
+      assert(pos <= max_n);
+      assert((max_n - pos) <= (SIZE_MAX / sizeof(Char)));
```
These should probably switch to be:
```cpp
CHECK_LE(pos, max_n);
CHECK_LE(max_n - pos, SIZE_MAX / sizeof(Char));
```
Same below.

dcposch replied:
@trevnorris @bnoordhuis fixed

trevnorris commented:
I'm fine with this.

@jasnell When are you cutting RC1?

Fishrock123 commented:
@trevnorris next week, probably tuesday

jasnell commented:
Planning to cut rc1 on Monday
On Apr 8, 2016 1:13 PM, "Trevor Norris" <notifications@github.com> wrote:

> I'm fine with this.
>
> @jasnell <https://github.com/jasnell> When are you cutting RC1?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/4846#issuecomment-207584207>
>

trevnorris commented:
@jasnell If this lands before Monday, could it make it in, or have all the RC commits been chosen?

jasnell commented:
Whatever makes it in before I start working on it on Monday ;)
On Apr 8, 2016 4:29 PM, "Trevor Norris" <notifications@github.com> wrote:

> @jasnell <https://github.com/jasnell> If this lands before Monday, could
> it make it in, or have all the RC commits been chosen?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/4846#issuecomment-207646415>
>

trevnorris commented:
Seems there's a minor performance hit with this PR, but I believe it's within an acceptable level. Anyone have any objections? If not then let's land it.

Fishrock123 commented:
🚢

jasnell commented:
Works for me.

jasnell commented:
@trevnorris ... there's still time to get this in. Is it ready to go?
@dcposch ... can you rebase?

dcposch commented:
> @dcposch ... can you rebase?

@jasnell yes, will be done v soon

dcposch commented:
@jasnell fixed

jasnell commented:
One final CI before landing: https://ci.nodejs.org/job/node-test-pull-request/2371/

dcposch commented:
@jasnell looks like `test-tls-inception` failed on FreeBSD and tests passed on the other platforms. I don't know if it's related to this change--looks unlikely. Want to try re-running it?

Failing build: https://ci.nodejs.org/job/node-test-commit-freebsd/2167/
Failing test: https://ci.nodejs.org/job/node-test-commit-freebsd/2167/nodes=freebsd10-64/tapTestReport

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/2372/

dcposch commented:
Sweet, everything worked that time including FreeBSD

jasnell commented:
Landed in 6c1e5ad

trevnorris commented:
@jasnell Was the squash/merge button used? I'm trying to figure out why the `Author:` field was changed from the listed commits.

jasnell commented:
No, I squashed like normal. Didn't notice that the author changed :-/
On Apr 25, 2016 3:48 PM, "Trevor Norris" <notifications@github.com> wrote:

> @jasnell <https://github.com/jasnell> Was the squash/merge button used?
> I'm trying to figure out why the Author: field was changed from the
> listed commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/4846#issuecomment-214554494>
>

trevnorris commented:
Strange. Oh well. Nothing serious.
